### PR TITLE
[BEAM-11546] Cross-language pipeline pattern

### PR DESCRIPTION
**Pull request by matthiasa4**
Pattern for cross-language pipeline pattern

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [x] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Dataflow | Flink | Samza | Spark | Twister2
--- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/) | ---
Java | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Twister2/lastCompletedBuild/)
Python | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python38/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/) | ---
XLang | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Direct/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Dataflow/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/) | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website | Whitespace | Typescript
--- | --- | --- | --- | --- | --- | ---
Non-portable | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocker_Cron/lastCompletedBuild/) <br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocs_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Whitespace_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Typescript_Cron/lastCompletedBuild/)
Portable | --- | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | --- | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by pabloem** [comment]
r: @rezarokni

**Comment by aromanenko-dev** [comment]
@matthiasa4 Hey Matthias, is it ready for review?

**Comment by matthiasa4** [comment]
@aromanenko-dev yes please! :)

**Comment by aromanenko-dev** [comment]
@matthiasa4 Could you rebase it with `master` branch? Since it has a conflict and the website rendering engine was significantly changed since last time.

**Comment by aromanenko-dev** [comment]
http://apache-beam-website-pull-requests.storage.googleapis.com/13642/documentation/patterns/cross-language/index.html
@matthiasa4 Hmm, interesting but python snippets seems not rendered properly and I see only the empty blocks. Is it me or you have the same issue?

**Comment by aromanenko-dev** [comment]
I believe we need to adapt code snippets to new rendering engine to which Beam website moved recently. 
Something like this:
```
{{< highlight java >}}
// Start by defining the options for the pipeline.
PipelineOptions options = PipelineOptionsFactory.create();

// Then create the pipeline.
Pipeline p = Pipeline.create(options);
{{< /highlight >}}

{{< highlight py >}}
Python code here
{{< /highlight >}}
```
Please, take a look on examples here:
https://github.com/apache/beam/blob/master/website/www/site/content/en/documentation/programming-guide.md

**Comment by matthiasa4** [comment]
Run Website_Stage_GCS PreCommit

**Comment by matthiasa4** [comment]
Run Website_Stage_GCS PreCommit

**Comment by matthiasa4** [comment]
Run Website_Stage_GCS PreCommit

**Comment by matthiasa4** [comment]
@aromanenko-dev think that did the trick! Thanks for pointing that out :)

File: website/www/site/content/en/documentation/patterns/cross-language.md [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+
+1. Start the **expansion service** with your Python transforms: `python expansion_service.py -p 9097`
+1. Start the **Job Server** which will translated into the stage that will run on your back-end or runner (e.g. Spark):
+
+   - From Apache Beam source code:
+     `./gradlew :runners:spark:job-server:runShadow`
+   - Using the pre-build Docker container:
+     `docker run -net=host apache/beam_spark_job_server`
+
+1. **Run pipeline**: `mvn exec:java -Dexec.mainClass=CrossLanguagePipeline -Pportable-runner -Dexec.args="--runner=PortableRunner --jobEndpoint=localhost:8099 --useExternal=true --expansionServiceURL=localhost:9097 --experiments=beam_fn_api"`
```
Comment (aromanenko-dev) [inline website/www/site/content/en/documentation/patterns/cross-language.md:167]: I'd suggest to put every option on the new line to make it more readable, something like:
```
mvn exec:java -Dexec.mainClass=CrossLanguagePipeline \
  -Pportable-runner \
  -Dexec.args=" \
    --runner=PortableRunner \
    --jobEndpoint=localhost:8099 \
    --useExternal=true \
    --expansionServiceURL=localhost:9097 \
    --experiments=beam_fn_api"
```

**Review by aromanenko-dev — APPROVED** [review]
Thanks! It LGTM in general from my technical point of view, just a minor note of formatting (ptal).
Also, perhaps it would make sense to reduce the padding size of code snippets to fit  code into the snippets frame or to break long lines into several ones. Now, there are several long code lines that are not easy to read because of this.

CC: @ibzib as one of the core Portability framework developers.

**Comment by matthiasa4** [comment]
Run Website_Stage_GCS PreCommit

**Comment by matthiasa4** [comment]
Run Website_Stage_GCS PreCommit

**Comment by aromanenko-dev** [comment]
@matthiasa4 Thanks, I think we are almost fine! 
Just one thing - please, don't merge a `master` branch into your feature branch. Instead, could you `rebase` your feature branch on top of `master` and also squash all your commits?

**Comment by matthiasa4** [comment]
Run Website_Stage_GCS PreCommit

**Comment by matthiasa4** [comment]
@aromanenko-dev Rebased & squashed, but some faults are flagged now - "file "code_samples/sdks_python_apache_beam_examples_snippets_snippets_test_py3.py" does not exist" - doesn't look like this is caused by my PR?

**Comment by aromanenko-dev** [comment]
Run Website PreCommit

**Comment by aromanenko-dev** [comment]
Run Website_Stage_GCS PreCommit

**Review by aromanenko-dev — APPROVED** [review]
It's green now, so LGTM

**Comment by matthiasa4** [comment]
Thanks for all the help @aromanenko-dev !

**Pull request merged**